### PR TITLE
Add support for dry-run

### PR DESCRIPTION
**Pull request by cPu1**
### Description

Adds dry-run support for `create cluster` and `create nodegroup`.

Closes https://github.com/weaveworks/eksctl/issues/3426

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Review by aclevername — CHANGES_REQUESTED** [review]
Looks good! Two things:
- Just out of curiosity have you tried using the config you get as an output as input to eksctl? It sounds silly but with the way we mutate the config I wouldn't be surprised if some bug exists/gets introduced meaning that it doesn't work. Could potentially be worth an integration test?
- I can see the integration test failing a lot as whenever we add a new field to the cluster/nodegroup struct you need to remember to update this test, which won't get noticed in a github PR as we don't run the integration tests. Could we move this test/copy a portion of this test to live somewhere that ensures it gets run as part of `make test`/PR flow?

**Comment by aclevername** [comment]
> It is! I was planning to address it in a separate PR, as part of adding an integration test for `create nodegroup`.

sweet!

> We don't run the integration tests for PRs, but we do build the integration tests, so any compilation errors will cause the PR checks to fail.

Its not compilation errors I'm worried about, its PRs like [this](https://github.com/weaveworks/eksctl/pull/2754) that are small so you wouldn't think to run the integration test, but would cause this test to fail as it introduces a new field. I think catching this before its merged would be really useful

**Comment by cPu1** [comment]
The dry-run integration test is failing because it needs AWS credentials. I'm working on it.

**Comment by aclevername** [comment]
> The dry-run integration test is failing because it needs AWS credentials. I'm working on it.

out of curiosity what sort of API calls do we make during dry-run that need creds? VPC validation?

**Comment by cPu1** [comment]
>out of curiosity what sort of API calls do we make during dry-run that need creds? VPC validation?

One of the first checks performed by every command is `sts.GetCallerIdentity`, so it's failing on that.

**Comment by cPu1** [comment]
Regarding running the dry-run integration test upon pushing to `main`, IMO, it's not worth the effort of introducing a new GH action that runs just the dry-run integration test to ensure any changes to the ClusterConfig schema are caught. I think we just have to accept that it's the same as other integration tests, and if the integration test suite didn't take too long to complete, we could've run it upon pushing to `main`. 

In the future, we can consider writing a custom check that reads the `defaultClusterConfig` variable and compares it with the output of the default ClusterConfig as part of the PR workflow.

File: pkg/ctl/create/cluster.go [hunk 1]
```diff
@@ -107,6 +110,15 @@ func doCreateCluster(cmd *cmdutils.Cmd, ngFilter *filter.NodeGroupFilter, params
...
 		return err
 	}
+
+	if params.DryRun {
+		originalWriter := logger.Writer
+		logger.Writer = io.Discard
+		defer func() {
+			logger.Writer = originalWriter
+		}()
+	}
```
Comment (aclevername) [inline pkg/ctl/create/cluster.go:120]: does this stop `logger.Error` calls from output?

**Review by aclevername — APPROVED** [review]
LGTM, just need to update the integration test to create a cluster using the output of the dry run

**Pull request merged**